tawazi/node/node.py

```diff
@@ -77,71 +78,89 @@ class ExecNode:
-        self.result: Union[NoValType, Any] = NoVal
-        """Internal attribute to store the result of the execution of this ExecNode (Might change!)."""
-        # even though setting result to NoVal is not necessary... it clarifies debugging
+    Raises:
+        ValueError: if setup and debug are both True.
+    """
 
-        self.profile = Profile(cfg.TAWAZI_PROFILE_ALL_NODES)
+    # 1. assign attributes
+    id_: Identifier
+    exec_function: Callable[..., Any] = field(default_factory=lambda: lambda *args, **kwargs: None)
+    priority: int = 0
+    is_sequential: bool = cfg.TAWAZI_IS_SEQUENTIAL
+    debug: bool = False
+    tag: Optional[TagOrTags] = None
+    setup: bool = False
+    unpack_to: Optional[int] = None
+    resource: Resource = cfg.TAWAZI_DEFAULT_RESOURCE
+
+    args: List["UsageExecNode"] = field(default_factory=list)  # args or []
+    kwargs: Dict[Identifier, "UsageExecNode"] = field(default_factory=dict)  # kwargs or {}
+
+    # TODO: fix _active behavior!
+    _active: Union[bool, "UsageExecNode"] = field(init=False, default=True)
+
+    # 4. Assign a default NoVal to the result of the execution of this ExecNode,
+    #  when this ExecNode will be executed, self.result will be overridden
+    # It would be amazing if we can remove self.result and make ExecNode immutable
+    result: Union[NoValType, Any] = field(init=False, default=NoVal)
+    """Internal attribute to store the result of the execution of this ExecNode (Might change!)."""
+    # even though setting result to NoVal is not necessary... it clarifies debugging
+
+    profile: Profile = field(default_factory=lambda: Profile(cfg.TAWAZI_PROFILE_ALL_NODES))
+
+    def __post_init__(self) -> None:
+        """Post init to validate attributes."""
+        # type verifications
+        is_none = self.tag is None
+        is_tag = isinstance(self.tag, Tag)
+        is_tuple_tag = isinstance(self.tag, tuple) and all(isinstance(v, Tag) for v in self.tag)
+        if not (is_none or is_tag or is_tuple_tag):
+            raise TypeError(
+                f"tag should be of type {TagOrTags} but {self.tag} of type {type(self.tag)} is provided"
+            )
+
+        if not isinstance(self.priority, int):
+            raise ValueError(f"priority must be an int, provided {type(self.priority)}")
+
+        if not isinstance(self.resource, Resource):
+            raise ValueError(f"resource must be of type {Resource}, provided {type(self.resource)}")
+
+        # other validations
+        if self.debug and self.setup:
+            raise ValueError(
+                f"The node {self.id} can't be a setup and a debug node at the same time."
+            )
+
+        if self.unpack_to is not None:
+            if not isinstance(self.unpack_to, int):
+                raise ValueError(
+                    f"unpack_to must be a positive int or None, provided {type(self.unpack_to)}"
+                )
+            # yes... empty tuples exist in Python
```

Comment:
```suggestion
            # empty tuple case
```
also  I think you can group the 2 ifs since the error msg is the same